### PR TITLE
Feature/trading

### PR DESCRIPTION
**Pull request by alex-stone**
### What changed? Why?

This adds trade support for address via:

```
address.trade(0.1, :eth, :usdc)
```

This utilizes the new trade APIs to construct a trade resource, server-side unsigned transaction(s), sign the transaction(s) within the SDK, and then broadcast it back using the server.

When performing ERC20 to ERC20 trades there are two transactions that need to be signed and broadcast, the first being an approve transaction and the latter being the transaction that will perform the trade.

This is only supported on `base-mainnet` currently as there is limited support for swaps on testnet (but this may be supported down the road).

#### Qualified Impact

This is a new feature

### CODE REVIEW

**Review by jazz-cb — APPROVED** [review]

File: lib/coinbase/transaction.rb [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+        data: parsed_payload['input'] || ''
+      }
+
+      @raw = Eth::Tx::Eip1559.new(Eth::Tx.validate_eip1559_params(params))
+    end
+
+    # Signs the Transaction with the provided key and returns the hex signing payload.
+    # @return [String] The hex-encoded signed payload
+    def sign(key)
+      raw.sign(key)
```
Comment (erdimaden) [inline lib/coinbase/transaction.rb:109]: @alex-stone should we update `@model.signed_payload` here?

Comment (alex-stone) [reply lib/coinbase/transaction.rb:109]: Comment left about this here: https://github.com/coinbase/coinbase-sdk-ruby/pull/73

I think that it makes sense to do that at some point given in the end-user custodial paradigm we will be broadcasting back the full transaction effectively.

I think we can iterate on that as we don't actually source the signed payload of this currently.

But yeah that's definitely an option!

**Review by erdimaden — APPROVED** [review]

File: lib/coinbase/trade.rb [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+
+        raise Timeout::Error, 'Trade timed out' if Time.now - start_time > timeout_seconds
+
+        self.sleep interval_seconds
+      end
+
+      self
+    end
+
+    # Reload reloads the Trade model with the latest version from the server side.
```
Comment (erdimaden) [inline lib/coinbase/trade.rb:110]:
```suggestion
    # Reloads the Trade model with the latest version from the server side.
```

File: spec/unit/coinbase/address_spec.rb [hunk 3]
```diff
@@ -449,6 +452,238 @@
...
+      end
+    end
+
+    describe 'when the to asset is unsupported' do
+      it 'raises an ArgumentError' do
+        expect do
+          address.trade(amount, from_asset_id, :uni)
+        end.to raise_error(ArgumentError, 'Unsupported to asset: uni')
+      end
+    end
```
Comment (erdimaden) [inline spec/unit/coinbase/address_spec.rb:662]: duplicated test

Comment (jazz-cb) [reply spec/unit/coinbase/address_spec.rb:662]: @erdimaden which are the duplicated tests?

File: lib/coinbase/trade.rb (hunk 2 above)
Comment (jazz-cb) [reply lib/coinbase/trade.rb]: fixed

File: spec/unit/coinbase/address_spec.rb (hunk 3 above)
Comment (erdimaden) [reply spec/unit/coinbase/address_spec.rb:662]: My bad, I misread that.

**Pull request merged**